Function Contracts Blogpost (#44)
model-checking/kani-verifier-blog: _posts/function-contracts.md

# Function Contracts for Kani

I this blogpost we discuss a new feature we’re developing for Kani: Function Contracts. It’s now available as an unstable feature with the `-Zfunction-contracts` flag. If you would like to learn more about the development and implementation details of this feature please refer to [the RFC](https://model-checking.github.io/kani/rfc/rfcs/0009-function-contracts.html). If you try out this new feature and want to leave feedback join the discussion in [the feature tracking issue](https://github.com/model-checking/kani/issues/2652).

## Introduction

> In today's blogpost we want to introduce you to a new concept that lets us verify larger programs: function contracts. Contracts let us safely break down the verification of a complex piece of code individually by function and efficiently compose the results, driving down the cost of verifying code with long call chains and repeated calls to the same function. This technique is called *modular verification* as the verification task is broken down into modules (in this case by function), verified independently and then recombined.

Reviewer: Sure

> The best example for how function contracts improve verification time are recursive functions. With a contract a recursive function can be verified in a single step, a technique called *inductive verification*. In this post we will explore how function contracts can be used to modularize the verification of a harness in the firecracker microvm by modularly verifying an implementation of Newton’s greatest common divisor algorithm inductively and then use that result to verify the harness.

Reviewer 2: ```suggestion
The best example for how function contracts improve verification time are recursive functions. With a contract a recursive function can be verified in a single step, a technique called *inductive verification*. In this post we will explore how function contracts can be used to modularize the verification of a harness in the [Firecracker](https://firecracker-microvm.github.io/) microvm by modularly verifying an implementation of Newton’s greatest common divisor algorithm inductively and then use that result to verify the harness.
```

## The Example: Firecracker’s `TokenBucket`

> We will explore employing a function contract when verifying the methods of `TokenBucket` in the `vmm` module of the firecracker microvm. To keep the example concise we will use the relatively simple `TokenBucket::new` function and its verification harness. The code is slightly simplified and we also use a recursive implementation of `gcd`. Firecracker actually uses a `gcd` with a loop which would need loop contracts to verify inductively. These are not supported yet by Kani, but conceptually they function the same way as function contracts. You can find the original version of all harnesses and verified code [here](https://github.com/firecracker-microvm/firecracker/blob/a774e26d63981fc031b974741a39519b08a61d3b/src/vmm/src/rate_limiter/mod.rs).

Reviewer 2: ```suggestion
We will explore employing a function contract when verifying the methods of `TokenBucket` in the `vmm` module of the Firecracker microvm. To keep the example concise we will use the relatively simple `TokenBucket::new` function and its verification harness. The code is slightly simplified and we also use a recursive implementation of `gcd`. Firecracker actually uses a `gcd` with a loop which would need loop contracts to verify inductively. These are not supported yet by Kani, but conceptually they operate in the same way as function contracts. You can find the original version of all harnesses and verified code [here](https://github.com/firecracker-microvm/firecracker/blob/a774e26d63981fc031b974741a39519b08a61d3b/src/vmm/src/rate_limiter/mod.rs).
```

> Lets first look at the harness that tests `TokenBucket::new`. It is straightforward, setting up a series of non-deterministic inputs, calling the function under verification and then asserting a validity condition.

Reviewer 2: ```suggestion
Lets first look at the harness that verifies `TokenBucket::new`. It is straightforward, setting up a series of non-deterministic inputs (using `kani::any_where` and `kani::any`), calling the function under verification and then asserting a validity condition.
```

```rust
pub fn new(size: u64, one_time_burst: u64, complete_refill_time_ms: u64) -> Option<Self> {
    if size == 0 || complete_refill_time_ms == 0 {
        return None;
    }
    let complete_refill_time_ns =
        complete_refill_time_ms.checked_mul(NANOSEC_IN_ONE_MILLISEC)?;
    let common_factor = gcd(size, complete_refill_time_ns);
    let processed_capacity: u64 = size / common_factor;
    let processed_refill_time: u64 = complete_refill_time_ns / common_factor;

    Some(TokenBucket {
        size,
        one_time_burst,
        initial_one_time_burst: one_time_burst,
        refill_time: complete_refill_time_ms,
        budget: size,
        last_update: Instant::now(),
        processed_capacity,
        processed_refill_time,
    })
}
```

Most of what happens here is just a series of divisions, with the exception of this call: `gcd(size, complete_refill_time_ns)` so let us look at that function next:

```rust
fn gcd(mut max: u64, mut min: u64) -> u64 {
    if min > max {
        std::mem::swap(&mut max, &mut min);
    }

    let rest = max % min;
    if rest == 0 { min } else { gcd(min, rest) }
}

```

> This is by far the costliest part of the code of `TokenBucket::new`. The rest of the function was a fixed set of divisions, but here we have a division and a recursive call back to `gcd`. It is not immediately obvious how busy this computation is but in the [worst case](https://en.wikipedia.org/wiki/Euclidean_algorithm#Worst-case) the number of steps (recursions) approaches 1.5 times the number of bits needed to represent the input numbers. Meaning that for two large 64 bit numbers it can take almost 96 iterations for a single call to `gcd`. Recall that our smallest input to `TokenBucket::new` is a non-deterministic value in the range $0< x < 18446744073709$ (up to 45 non-zero bits). If a harness that uses `gcd` does not heavily constrain the input, Kani would have to unroll the recursion at least 68 times and then execute it symbolically. An expensive operation.

Reviewer 2: ```suggestion
This is by far the costliest part of the code of `TokenBucket::new`. The rest of the function was a fixed set of divisions, but here we have a division and a recursive call back to `gcd`. It is not immediately obvious how busy this computation is but in the [worst case](https://en.wikipedia.org/wiki/Euclidean_algorithm#Worst-case) the number of steps (recursions) approaches 1.5 times the number of bits needed to represent the input numbers. Meaning that for two large 64 bit numbers it can take almost 96 iterations for a single call to `gcd`. Recall that our smallest input to `TokenBucket::new` is a non-deterministic value in the range $0< x < 18446744073709$ (up to 45 non-zero bits). If a harness that uses `gcd` does not heavily constrain the input, Kani would have to unroll the recursion at least 68 times and then execute it symbolically. An expensive operation.
```

Reviewer 3: @roypat If you have time, do you think you could help us by reviewing this post? Thank you!

Reviewer 4: Hi @celinval, sure I can have a look! I might only get some time next week though :(

Reviewer 2: I have no idea either now 😅 I guess we can leave it.

> Since `gcd` is the most expensive part of verifying this harness it is an ideal target for using a contract to make it more efficient. In fact by replacing it with a contract we not only make this harness more efficient but in fact any harness that uses `gcd` directly or indirectly (for instance by calling `TokenBucket::new`. In firecracker for instance there are three more harnesses that all call `gcd` and thus benefit from modular verification.

Reviewer 2: ```suggestion
Since `gcd` is the most expensive part of verifying this harness it is an ideal target for using a contract to make it more efficient. In fact by replacing it with a contract we not only make this harness more efficient but also any harness that uses `gcd` directly or indirectly (e.g., by calling `TokenBucket::new`). In Firecracker, for instance, there are three more harnesses that all call `gcd` and thus benefit from modular verification.
```

## Introducing a Function Contract

Function contracts are conceptually rather simple actually. They comprise a set of conditions which characterize the behavior of the function, similar to the conditions you would use in a test case. There are different types of conditions which are also called the *clauses* of the contract. The first type of clause we will introduce here is the `ensures` clause.

```rust
#[kani::ensures(max % result == 0 && min % result == 0 && result != 0)]
fn gcd(mut max: u64, mut min: u64) -> u64 {
    if min > max {
        std::mem::swap(&mut max, &mut min);
    }

    let rest = max % min;
    if rest == 0 { min } else { gcd(min, rest) }
}

```

> The `ensures` clause describes the relationship of the return of the function with the arguments that the function was called with. It is often also called a *postcondition*, because it is a *condition* that must hold after (*post*) the execution of the function. With Kani the contents of the `ensures` clause can be any Rust expression that returns `bool`. However the expression may not allocate, deallocate or modify heap memory or perform I/O. A single function may have multiple `ensures` clauses which functions as though they had ben joined with `&&`. Our example could thus also have been written as

Reviewer 2: ```suggestion
The `ensures` clause describes the relationship of the return of the function (i.e., `result`) with the arguments that the function was called with. It is often also called a *postcondition*, because it is a *condition* that must hold after (*post*) the execution of the function. With Kani the contents of the `ensures` clause can be any Rust expression that returns `bool`. However the expression may not allocate, deallocate or modify heap memory or perform I/O. A single function may have multiple `ensures` clauses which functions as though they had been joined with `&&`. Our example could thus also have been written as
```

To put it another way, when we recurse, we are allowed to assume that the verification passed, because the step that we are currently verifying is being checked for any possible input, including the one we are recursing with. If a problem were to occur anywhere in the recursive calls, we would see the same problem as a verification failure of the first step.

> Notice that using this trick, we reduced the up to 64 unrollings to just one, a substantial win. What is even better that doing this requires no input from you. Kani does this automatically. Any function with a contract is always verified inductively. Side note: this works even if the recursive call is not direct but buried somewhere in the call chain.

Reviewer 2: I meant for contracts.

> Notice that using this trick, we reduced the up to 64 unrollings to just one, a substantial win. What is even better that doing this requires no input from you. Kani does this automatically. Any function with a contract is always verified inductively. Side note: this works even if the recursive call is not direct but buried somewhere in the call chain.

Author: I can only think of one which is and isn't specific to inductive verification. Technically using induction requires that the return type implement `Arbitrary` because it want's to use the replacement. Normally a contract check does not require `Arbitrary`, only the replacement does. However there is currently no way to turn off generation of the replacement function, and so if the return type didn't implement `Arbitrary` it would fail to compile regardless of whether we use inductive verification or not.

> Notice that using this trick, we reduced the up to 64 unrollings to just one, a substantial win. What is even better that doing this requires no input from you. Kani does this automatically. Any function with a contract is always verified inductively. Side note: this works even if the recursive call is not direct but buried somewhere in the call chain.

Reviewer 2: I don't think we need to add this to the text at this point! Thanks!

    // Create non-determinstic values
    let max = kani::any();
    let min = kani::any();

    gcd(max, min)
}

```

This may be surprising to you since, in our example, we use a straightforward `kani::any()`. Indeed for finite types like `u64` that implement `Arbitrary` Kani *could* generate this harness automatically. However for any type involving references or pointers (like `Vec`) Kani is unable to generate the harness. Because it would be confusing if sometimes the harness is generated automatically and sometimes not, we decided that for the time being a manual harness is always required. We are working on adding auto generation in the future.

A few words about writing contract harnesses: The harness should generate completely unconstrained values, otherwise the verification will be unsound. Usually this means calling `kani::any()`. Sometimes it is not possible to create completely unconstrained values, as is the case with recursive types and array-backed types. In these cases you must use careful judgement and ensure that the value you create is large enough to exercise all possible behaviors of the function. You can think of this as covering all branches.

> Finally I told you we would be adding a check to ensure that our `gcd` actually produces the greatest divisor. For this we will need a separate harness and in this case we sadly cannot use any contract substitution. The reason is that, if we substitute, we get values that satisfy the postcondition, but nothing more. Since the postcondition does not ensure the result is the largest divisor possible, it won’t be.

Reviewer: Do you mean Kani unwinds the recursion all the way?

> Finally I told you we would be adding a check to ensure that our `gcd` actually produces the greatest divisor. For this we will need a separate harness and in this case we sadly cannot use any contract substitution. The reason is that, if we substitute, we get values that satisfy the postcondition, but nothing more. Since the postcondition does not ensure the result is the largest divisor possible, it won’t be.

Author: Yes. As you rightly point out that's the only way to make this proof work.

# Function Contracts for Kani

In this blogpost we discuss function contracts which are now available as an unstable feature, enabled with the `-Zfunction-contracts` flag. If you would like to learn more about the development and implementation details of this feature please refer to [the RFC](https://model-checking.github.io/kani/rfc/rfcs/0009-function-contracts.html). If you try out this new feature and want to leave feedback join the discussion in [the feature tracking issue](https://github.com/model-checking/kani/issues/2652).

## Introduction

Today we want to introduce you to a new feature in Kani that lets us verify larger programs: function contracts. Contracts let us safely break down the verification of a complex piece of code individually by function and efficiently compose the results, driving down the cost of verifying code with long call chains and repeated calls to the same function. This technique is called *modular verification* as the verification task is broken down into modules (in this case by function), verified independently and then recombined.

> The best example for how function contracts improve verification time are recursive functions. With a contract a recursive function can be verified in a single step, a technique called *inductive verification*. In this post we will explore how function contracts can be used to modularize the verification of a harness in the [Firecracker](https://firecracker-microvm.github.io/) microVM by modularly verifying an implementation of Euclid’s greatest common divisor algorithm inductively and then use that result to verify the harness.

Reviewer 4: ```suggestion
The best example for how function contracts improve verification time are recursive functions. With a contract a recursive function can be verified in a single step, a technique called *inductive verification*. In this post we will explore how function contracts can be used to modularize the verification of a harness in the [Firecracker](https://firecracker-microvm.github.io/) Virtual Machine Monitor by modularly verifying an implementation of Euclid’s greatest common divisor algorithm inductively and then use that result to verify the harness.
```

## The Example: Firecracker’s `TokenBucket`

> We will explore employing a function contract when verifying the methods of `TokenBucket` in the `vmm` module of the Firecracker microVM. To keep the example concise we will use the relatively simple `TokenBucket::new` function and its verification harness. The code is slightly simplified and we also use a recursive implementation of `gcd`. Firecracker actually uses a `gcd` with a loop which would need loop contracts to verify inductively. These are not supported yet by Kani, but conceptually they function the same way as function contracts. You can find the original version of all harnesses and verified code [here](https://github.com/firecracker-microvm/firecracker/blob/a774e26d63981fc031b974741a39519b08a61d3b/src/vmm/src/rate_limiter/mod.rs).

Reviewer 4: Firecracker is not a microVM itself, its a virtual machine monitor (e.g. it manages a microVM)

```suggestion
We will explore employing a function contract when verifying the methods of `TokenBucket` in the `vmm` module of the Firecracker Virtual Machine Manager. To keep the example concise we will use the relatively simple `TokenBucket::new` function and its verification harness. The code is slightly simplified and we also use a recursive implementation of `gcd`. Firecracker actually uses a `gcd` with a loop which would need loop contracts to verify inductively. These are not supported yet by Kani, but conceptually they function the same way as function contracts. You can find the original version of all harnesses and verified code [here](https://github.com/firecracker-microvm/firecracker/blob/a774e26d63981fc031b974741a39519b08a61d3b/src/vmm/src/rate_limiter/mod.rs).
```

Also, maybe we could link to the previous blog post about Firecracker here? (totally not biased because I'm wrote that one xP)

```rust
pub fn new(size: u64, one_time_burst: u64, complete_refill_time_ms: u64) -> Option<Self> {
    if size == 0 || complete_refill_time_ms == 0 {
        return None;
    }
    let complete_refill_time_ns =
        complete_refill_time_ms.checked_mul(NANOSEC_IN_ONE_MILLISEC)?;
    let common_factor = gcd(size, complete_refill_time_ns);
    let processed_capacity: u64 = size / common_factor;
    let processed_refill_time: u64 = complete_refill_time_ns / common_factor;

    Some(TokenBucket {
        size,
        one_time_burst,
        initial_one_time_burst: one_time_burst,
        refill_time: complete_refill_time_ms,
        budget: size,
        last_update: Instant::now(),
        processed_capacity,
        processed_refill_time,
    })
}
```

Most of what happens here is just a series of divisions, with the exception of this call: `gcd(size, complete_refill_time_ns)` so let us look at that function next:

```rust
fn gcd(mut max: u64, mut min: u64) -> u64 {
    if min > max {
        std::mem::swap(&mut max, &mut min);
    }

    let rest = max % min;
    if rest == 0 { min } else { gcd(min, rest) }
}

> 

Reviewer 2: Remove extra line.

```rust
#[kani::ensures(max % result == 0 && min % result == 0 && result != 0)]
fn gcd(mut max: u64, mut min: u64) -> u64 {
    if min > max {
        std::mem::swap(&mut max, &mut min);
    }

    let rest = max % min;
    if rest == 0 { min } else { gcd(min, rest) }
}

```

The `ensures` clause describes the relationship of the return of the function (e.g. `result`) with the arguments that the function was called with. It is often also called a *postcondition*, because it is a *condition* that must hold after (*post*) the execution of the function. With Kani, the contents of the `ensures` clause can be any Rust expression that returns `bool`. However the expression may not perform any *side effects*, that is: allocate, deallocate or modify heap memory or perform I/O. A single function may have multiple `ensures` clauses which functions as though they had been joined with `&&`. Our example could thus also have been written as

```rust
#[kani::ensures(max % result == 0)]
#[kani::ensures(min % result == 0)]
#[kani::ensures(result != 0)]
fn gcd(mut max: u64, mut min: u64) -> u64 { ... }
```

> You may wonder at this point what this is actually useful for. It turns out that for a verifier the *abstraction* of the function as described by the conditions in our `ensures` clause, is much easier to reason about than running through the many recursive calls in the actual implementation of `gcd`. You may also notice that the abstraction only approximates `gcd`. `max % result == 0` and `min % result == 0` describe *a* common divisor of `max` and `min` but not necessarily the *greatest*. However in this case this is acceptable to us when it comes to verifying the callers of `gcd`. Later we will add an additional check to ensures we generate the largest divisor.

Reviewer: ```suggestion
You may wonder at this point what this is actually useful for. It turns out that for a verifier the *abstraction* of the function as described by the conditions in our `ensures` clause, is much easier to reason about than running through the many recursive calls in the actual implementation of `gcd`. You may also notice that the abstraction only approximates `gcd`. `max % result == 0` and `min % result == 0` describe *a* common divisor of `max` and `min` but not necessarily the *greatest*. However in this case this is acceptable to us when it comes to verifying the callers of `gcd`. Later we will add an additional check to ensure we generate the largest divisor.
```
OR "that ensures"

```rust
#[kani::requires(max != 0 && min != 0)]
#[kani::ensures(max % result == 0 && min % result == 0 && result != 0)]
fn gcd(mut max: u64, mut min: u64) -> u64 { ... }
```

As with postconditions, the precondition allows any side-effect free rust expressions of type `bool`, may mention the function arguments and multiple `requires` clauses act as though they were joined with `&&`.

To understand how this makes our contract verification succeed, let's integrate it into the manual harness we had written before.

```rust
#[kani::proof]
fn gcd_stub_check() {
    // First we create any possible input using a
    // non-deterministic value
    let max = kani::any();
    let min = kani::any();

    // Limit the domain of inputs with precondition
    kani::assume(max != 0 && min != 0);

    // We create any possible result
    let result = gcd(max, min);

    assert!(
        // And here is our postcondition
        max % result == 0 && min % result == 0 && result != 0
    );
}

> 

Reviewer 2: Remove extra line.

As a small technicality: induction is usually split into the so called "base case" and the "induction step". In our example a single run of the function actually combines both, which is the `if rest == 0` split. When `rest == 0` then we have reached the base case, otherwise we are in the induction step. Both the base case and the induction step must uphold the ponstconditions. The way that our proof is set up here already ensures that because the postcondition `assert!` gets enforced regardless of whether we took the base case or the induction step.[^return] 

> [^return]: The astute reader might note that this would not be the case if the body of the function we are verifying contained a `return` statement. Rest assured that the code you see here is jsut an illustration and the actual code generated for contract verification is not so easily fooled. If you want to see actual examples of how this is set up there is one in the implementation documentation of the [contracts macros](https://github.com/model-checking/kani/blob/main/library/kani_macros/src/sysroot/contracts.rs) or you can dump the generated code with the rustc flag `-Zunpretty=expanded`.

Reviewer 2: ```suggestion
[^return]: The astute reader might note that this would not be the case if the body of the function we are verifying contained a `return` statement. Rest assured that the code you see here is just an illustration and the actual code generated for contract verification is not so easily fooled. If you want to see actual examples of how this is set up there is one in the implementation documentation of the [contracts macros](https://github.com/model-checking/kani/blob/main/library/kani_macros/src/sysroot/contracts.rs) or you can dump the generated code with the rustc flag `-Zunpretty=expanded`.
```

To put it another way, when we recurse, we are allowed to assume that the verification passed, because the step that we are currently verifying is being checked for any possible input, including the one we are recursing with. If a problem were to occur anywhere in the recursive calls, we would see the same problem as a verification failure of the first step.

> Notice that using this trick, we reduced the up to 64 unrollings to just one, a substantial win. What is even better that doing this requires no input from you. Kani does this automatically. Any function with a contract is always verified inductively. Side note: this works even if the recursive call is not direct but buried somewhere in the call chain.

Reviewer 2: Wasn't "up to 68 unrollings" as you previously mentioned?

> ## A bit of Cleanup

Reviewer 2: ```suggestion
## A Bit of Cleanup
```

Finally we mentioned we would be adding a check to ensure that our `gcd` actually produces the greatest divisor. For this we will need a separate harness and in this case we sadly cannot use any contract substitution. The reason is that, if we substitute, we get values that satisfy the postcondition, but nothing more. Since the postcondition does not ensure the result is the largest divisor possible, it won’t be.

```rust
#[kani::proof]
fn gcd_greatest_check() {
    // Create non-determinstic values
    let max = kani::any();
    let min = kani::any();

    let result = gcd(max, min);

    let greatest = kani::any_where(|a|
        max % *a == 0 && min % *a == 0
    );
    assert!(!(greatest > result));
}

```

> Currently our function contracts can’t express that the result is the largest divisor, which necessitates this additional harness. In future Kani’s function contracts will be extended with *quantifiers*, which will allow the postcondition to express this property. For instance such a postcondition may look like this: `forall(|i| : max % i == 0 && min % i == 0 => i <= result)` which states that if any integer `i` exists that is also a divisor of `min` and `max`, then it must either be the result or smaller than it. Clearly such a condition can only be satisfied by a `result` that is the largest common divisor.

Reviewer 4: :heart_eyes:

## Soundness and Comparison with Stubbing

Function contracts and stubbing are closely related as both allow the replacement of a costly computation with a cheap one via harness attributes. The crucial difference is that contracts preserve soundness. We will get into more detail later, but put simply: function contracts being sound means you can trust a successful verification.

Conceptually stub `#[kani::stub(target, source)]` replaces all calls to `target` with `source` with weak constraints on what `source` is. `source` needs to be a function definition with a signature such that each call after the replacement typechecks. The details of what that means are unimportant, what matters is that no behavioral equivalence is checked beyond type compatibility.

To illustrate this, lets consider an [example from XKCD](https://xkcd.com/221/) which stubs a `random` function by one that just returns `4`. This type of use of stubbing is completely legal and would not lead to a type error or verification failure.

```rust
// extern
crate rand {
    pub fn random<T>() -> T
    where
        Standard: Distribution<T>;

> }

Reviewer 4: Mh, I get what you want to express here, but this is not really valid Ruts syntax. Maybe its better to just say `extern crate rand` with a comment that it exposes the `random` function?

Author: I've replaced with an empty/opaque body. I want the type signature in there to show that stubs don't even need to have the same trait bounds or types as what they replace so long as it still works with how the function actually ends up being called.

> fn getRandomNumber() -> u32 {

Reviewer 4: Here I would change the function to snake case, even if the xkcd uses camel case

    4  // chosen by fair dice roll
       // guaranteed to be random
}

#[kani::proof]
#[kani::stub(rand::random, getRandomNumber)]
fn my_program() {
    let v = vec![1;5];
    let i : u32 = rand::random();

    println!("{}", v[i as usize]);
}
```

> It is rather obvious that `getRandomNumer` is not the same as `rand::random`. In the harness included in the example we can see how this leads to problems. After the stubbing is applied the random number is always going to be 4, which means the verification of the harness will succeed. However at runtime, where `rand::random` is no longer replaced a number greater than 4 can be produced and cause a panic. This type of situation where a runtime panic is not caught by verification is referred to as a *violation of* *soundness* or *unsoundness*.

Reviewer 4: ```suggestion
It is rather obvious that `getRandomNumber` is not the same as `rand::random`. In the harness included in the example we can see how this leads to problems. After the stubbing is applied the random number is always going to be 4, which means the verification of the harness will succeed. However at runtime, where `rand::random` is no longer replaced a number greater than 4 can be produced and cause a panic. This type of situation where a runtime panic is not caught by verification is referred to as a *violation of* *soundness* or *unsoundness*.
```

When *soundness* is preserved that means that any runtime problem (such as a panic) cause a verification failure. Dually if soundness is violated then verification succeeds, but a runtime error is still possible. As you may imagine unsoundness significantly weakens the usefulness of a verifier. After all why bother verifying if we can still get issues at runtime? As a result verifiers like Kani take great care to preserve soundness. Stubbing is one of the features that ***can*** introduce unsoundness. As a result it is important to ensure that, at least for the code under verification, any stub behaves like the function it replaces. For instance it is completely sound to replace `rand::random` with a non-deterministic value with the following implementation.

```rust
fn getRandomNumber() -> u32 {
    kani::any() // actually all possible values
                // guaranteed to be random
}
```

> The verifier cannot execute the actual `rand::random`, because that relies on a syscall that returns a source of randomness from the machine, and so we need to replace it with something for the purposes of verification and since non-deterministic values are actually equivalent to random values this use of stubbing is useful, safe and necessary.

Reviewer 2: I know what you mean, but I don't think it's appropriate to say that "non-deterministic values are equivalent to random values". I'd prefer to say "since a non-deterministic value will cover all possible values produced by `rand::random` this use of..."

Author: I've rephrased this let me know if that's better.

Reviewer 2: Way better, thanks!

> You may wonder how we can determine whether a given situation is sound. We do this by inspecting the relationship between the a function and its replacement. The replacement, by design, always approximates the behavior of the code it replaces and the question is how it does this. Conceptualize the original function as simply a black box that produces a set of possible outputs. If it’s replacement produces a subset of the outputs then that is called an *underapproximation*, which causes unsoundness. We see this in our example where `getRandomNumber` produced fewer (e.g. only the integer `4`) instead of all possible outputs (all `u32` integers). Conversely if the replacement produces a superset of outputs this is called an *overapproximation* and crucially always sound!

Reviewer 2: ```suggestion
You may wonder how we can determine whether a given situation is sound. We do this by inspecting the relationship between a function and its replacement. The replacement, by design, always approximates the behavior of the code it replaces and the question is how it does this. Conceptualize the original function as simply a black box that produces a set of possible outputs. If its replacement produces a subset of the outputs, then that is called an *under-approximation*, which causes unsoundness. We see this in our example where `getRandomNumber` produced fewer (e.g. only the integer `4`) instead of all possible outputs (all `u32` integers). Conversely if the replacement produces a superset of outputs this is called an *over-approximation* and crucially always sound!
```

Side note: the model of the black box function also extends to functions that take arguments, though in this case we have to consider separately the output sets for every combination of argument values.

> Function contracts are designed such that they are always overapproximations and thus sound. This is guaranteed by the following: any value the implementation produces must pass the postcondition, but not every value that passes the postcondition must be produced by the implementation. During replacement with the contract all values that pass the postcondition are created, thus a superset of the actual outputs.

Reviewer 2: ```suggestion
Function contracts are designed such that they are always over-approximations and thus sound. This is guaranteed by the following: any value the implementation produces must pass the postcondition, but not every value that passes the postcondition must be produced by the implementation. During replacement with the contract all values that pass the postcondition are created, thus a superset of the actual outputs.
```

## Introduction

Today we want to introduce you to a new feature in Kani that lets us verify larger programs: function contracts. Contracts let us safely break down the verification of a complex piece of code individually by function and efficiently compose the results, driving down the cost of verifying code with long call chains and repeated calls to the same function. This technique is called *modular verification* as the verification task is broken down into modules (in this case by function), verified independently and then recombined.

The best example for how function contracts improve verification time are recursive functions. With a contract a recursive function can be verified in a single step, a technique called *inductive verification*. In this post we will explore how function contracts can be used to modularize the verification of a harness in the [Firecracker](https://firecracker-microvm.github.io/) Virtual Machine Monitor by modularly verifying an implementation of Euclid’s greatest common divisor algorithm inductively and then use that result to verify the harness.

> Aside: we've actually blogged twice previously about the exiting verification work we've been doing on the Firecracker project. You can find those posts [here]({% link _posts/2023-08-31-using-kani-to-validate-security-boundaries-in-aws-firecracker.md %}) and [here]({% link _posts/2022-07-13-using-the-kani-rust-verifier-on-a-firecracker-example.md %}).

Reviewer 4: ```suggestion
Aside: we've actually blogged twice previously about the exciting verification work we've been doing on the Firecracker project. You can find those posts [here]({% link _posts/2023-08-31-using-kani-to-validate-security-boundaries-in-aws-firecracker.md %}) and [here]({% link _posts/2022-07-13-using-the-kani-rust-verifier-on-a-firecracker-example.md %}).
```

Our goal will be to eventually use the efficient abstraction everywhere where `gcd` is called but first we must ensure that `gcd` respects these conditions. We need to *verify the contract*.

Doing so is the same as verifying that the postcondition(s) hold for any possible input to the function. If we were to manually create a harness that does this verification it would look like this

```rust
#[kani::proof]
fn gcd_stub_check() {
    // First we create any possible input using a
    // non-deterministic value
    let max = kani::any();
    let min = kani::any();

    // We create any possible result
    let result = gcd(max, min);

    assert!(
        // And here is our postcondition
        max % result == 0 && min % result == 0 && result != 0
    );
}

> 

Reviewer 2: Remove extra line.

Author: I'm sorry these must have reappeared after a merge. I thought we'd removed all of them.

---

## A Neat Trick: Contracts for Non-local Functions

Attributes for attaching function contracts only work on crate-local items but you may wish to stub an out of crate function. There is currently no builtin way to do so but you can achieve the same effect using a technique known as the "double stub". The idea is to first stub an external function to a local function that immediately calls the external function. Then the desired contract is attached to this local function which is then then stubbed as a `stub_verified` using the contract.

```rust
use external_crate::gcd;

#[kani::ensures(result < max && result < min && result != 0)]
fn local_gcd(max: i32, min: i32) -> i32 {
    // immediate call with same arguments to the external function
    gcd(max, min)
}

fn function_under_verification(...) {
    ... gcd(...) ...
}

#[kani::proof]
#[kani::stub(gcd, local_gcd)]
#[kani::stub_verified(local_gcd)]
fn harness() {
    ... function_under_verification(...) ...
}

> 

Reviewer 2: Remove extra line.
